source/served/commands/complete.d

```diff
@@ -924,21 +935,131 @@ unittest
 	assert(!isInComment("int x;\n// line comment\n", 23, backend));
 }
 
-auto convertDCDIdentifiers(DCDIdentifier[] identifiers, bool argumentSnippets, bool completeNoDupes)
+auto convertDCDIdentifiers(DCDIdentifier[] identifiers, bool argumentSnippets, bool completeNoDupes, DCDExtComponent dcdext)
 {
 	CompletionItem[] completion;
 	foreach (identifier; identifiers)
 	{
 		CompletionItem item;
+		string detailDetail, detailDescription;
 		item.label = identifier.identifier;
 		item.kind = identifier.type.convertFromDCDType;
 		if (identifier.documentation.length)
 			item.documentation = MarkupContent(identifier.documentation.ddocToMarked);
-		if (identifier.definition.length)
+		
+		if (identifier.definition.length == 0)
+		{
+			if (identifier.type.length == 1)
+			{
+				switch (identifier.type[0])
+				{
+				case 'c':
+					detailDescription = "Class";
+					break;
+				case 'i':
+					detailDescription = "Interface";
+					break;
+				case 's':
+					detailDescription = "Struct";
+					break;
+				case 'u':
+					detailDescription = "Union";
+					break;
+				case 'a':
+					detailDescription = "Array";
+					break;
+				case 'A':
+					detailDescription = "AA";
+					break;
+				case 'v':
+					detailDescription = "Variable";
+					break;
+				case 'm':
+					detailDescription = "Member";
+					break;
+				case 'e':
+					// lowercare to differentiate member from enum name
+					detailDescription = "enum";
+					break;
+				case 'k':
+					detailDescription = "Keyword";
+					break;
+				case 'f':
+					detailDescription = "Function";
+					break;
+				case 'g':
+					detailDescription = "Enum";
+					break;
+				case 'P':
+					detailDescription = "Package";
+					break;
+				case 'M':
+					detailDescription = "Module";
+					break;
+				case 't':
+				case 'T':
+					detailDescription = "Template";
+					break;
+				case 'h':
+					detailDescription = "<T>";
+					break;
+				case 'p':
+					detailDescription = "<T...>";
+					break;
+				case 'l': // Alias (eventually should show what it aliases to)
+				default:
+					break;
+				}
+			}
+		}
+		else
 		{
 			item.detail = identifier.definition;
-			if (!completeNoDupes)
-				item.sortText = identifier.definition;
+
+			// check if that's actually a proper completion item to process
+			auto definitionSpace = identifier.definition.indexOf(' ');
+			if (definitionSpace != -1)
+			{
+				detailDescription = identifier.definition[0 .. definitionSpace];
```

Comment:
this should have a comment explaining a few cases what this does / how this looks (as just dropping everything after space could be suboptimal)